### PR TITLE
More category fixes

### PR DESCRIPTION
**Pull request by thecaligarmo**
This fix does two things:

1. Adds `category.subcategores` which is a `list` of categories which contain `category` as a parent.
2. Updates plugin to use the latest code testing version for pelican-plugins.

### CODE REVIEW

**Comment by oulenz** [comment]
Thanks for your contribution!

I think it's best if we think through the various use cases. I assume you need access to all the ancestors of a category, not just its children?

If so, I think we should seperately offer access to the children of a category.

**Comment by thecaligarmo** [comment]
There is already functionality to get the ancestors (parents) of a category. It did not have the functionality to get the children. So what I needed was access to the children.

What you're suggesting is exactly what I added. I separately added access to the children of a category.

**Comment by oulenz** [comment]
Sorry, I meant to say descendents. You are adding all the descendents, but I could image someone wanting to list only the children. (E.g. the way the content of categories is displayed on Wikipedia.)

File: README.md [hunk 1]
```diff
@@ -20,6 +20,13 @@ or not, use:
         <a href="{{ SITEURL }}/{{ cat.url }}">{{ cat }}</a>{{ ', ' if not loop.last }}
     {% endfor %}
 
+Additionally, this plugin adds `category.subcategories`: a `list` of categories
+that have `category` as a parent.
+
+    {% for subcat in category.subcategories %}
+        <a href="{{ SITEURL }}/{{subcat.url}}">{{subcat.shortname|capitalize}}</a>
+    {% endfor %}
+
```
Comment (oulenz) [inline README.md]: This should be in the section below (Nested Categories), towards the end

File: pelican/plugins/more_categories/more_categories.py [hunk 2]
```diff
@@ -3,7 +3,7 @@
 Title: More Categories
 Description: adds support for multiple categories per article and nested
 categories
-Requirements: Pelican 3.8 or higher
+Requirements: Pelican 4.2.0 or higher
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: Why 4.2.0? (3.8 became 4.0, but that should be enough.)

File: pelican/plugins/more_categories/more_categories.py [hunk 3]
```diff
@@ -19,13 +19,13 @@ class Category(URLWrapper):
     @property
     def _name(self):
         if self.parent:
-            return self.parent._name + '/' + self.shortname
+            return self.parent._name + "/" + self.shortname
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: Please don't change single quotes to double quotes.

File: pelican/plugins/more_categories/more_categories.py [hunk 4]
```diff
@@ -38,14 +38,13 @@ def name(self, val):
...
         if self._slug is None:
-            if 'CATEGORY_REGEX_SUBSTITUTIONS' in self.settings:
-                subs = self.settings['CATEGORY_REGEX_SUBSTITUTIONS']
+            if "CATEGORY_REGEX_SUBSTITUTIONS" in self.settings:
+                subs = self.settings["CATEGORY_REGEX_SUBSTITUTIONS"]
             else:
-                subs = self.settings.get('SLUG_REGEX_SUBSTITUTIONS', [])
+                subs = self.settings.get("SLUG_REGEX_SUBSTITUTIONS", [])
             self._slug = slugify(self.shortname, regex_subs=subs)
-            print(self._slug)
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py:46]: Good catch!

File: pyproject.toml [hunk 5]
```diff
@@ -46,7 +46,7 @@ invoke = "^1.3"
 isort = "^4.3"
 livereload = "^2.6"
 markdown = "^3.1.1"
-pytest = "^5.0"
+pytest = "5.3.5"
```
Comment (oulenz) [inline pyproject.toml]: Why did you change this?

File: tasks.py [hunk 6]
```diff
@@ -1,34 +1,93 @@
...
-FLAKE8 = 'flake8'
-ISORT = 'isort'
-PYTEST = 'pytest'
+PKG_NAME = "more_categories"
+PKG_PATH = Path(f"pelican/plugins/{PKG_NAME}")
+
+ACTIVE_VENV = os.environ.get("VIRTUAL_ENV", None)
+VENV_HOME = Path(os.environ.get("WORKON_HOME", "~/.local/share/virtualenvs"))
+VENV_PATH = Path(ACTIVE_VENV) if ACTIVE_VENV else (VENV_HOME / PKG_NAME)
+VENV = str(VENV_PATH.expanduser())
```
Comment (oulenz) [inline tasks.py:13]: I prefer activating a suitable virtual environment manually. In particular because I (and maybe others) don't use `virtualenv` but `conda`.

File: tasks.py [hunk 7]
```diff
@@ -1,34 +1,93 @@
...
+def black(c, check=False, diff=False):
+    """Run Black auto-formatter, optionally with --check or --diff"""
+    check_flag, diff_flag = "", ""
     if check:
-        check_flag = '-c'
-    c.run(f'{ISORT} {check_flag} --recursive {PKG_PATH}/* tasks.py')
+        check_flag = "--check"
+    if diff:
+        diff_flag = "--diff"
+    c.run(f"{VENV}/bin/black {check_flag} {diff_flag} {PKG_PATH} tasks.py")
```
Comment (oulenz) [inline tasks.py]: This repo doesn't use black.

File: RELEASE.md [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Release type: minor
+
+Added subcategories to categories to be able to cycle through which subcategories are present for a category.
+Updated plugin to better work with modern pelican standards (by updating tasks.py and ensuring proper code formatting)
```
Comment (oulenz) [inline RELEASE.md:4]: I should probably add a changelog, but I don't understand this addition. Is it it meant to be converted to a changelog by some automation tool?

**Review by oulenz — CHANGES_REQUESTED** [review]
I've added some inline comments. I think this PR would be easier to handle if you just proposed the descendents/children functionality (and not all the code style changes).

File: pelican/plugins/more_categories/more_categories.py (hunk 2 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: That should say 4.0.0. Sorry. The reason I changed it is because in the README it said 4.0.0. From what I can tell, it's requiring 4.0 functionality, so I decided to change it. Was that wrong?

File: pelican/plugins/more_categories/more_categories.py (hunk 3 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: This is done automatically using the tasks that are present within pelican. In the lint section they are using "black". Black makes it so that single quotes are turned to double quotes to make coding conventions the same across the board. Since black is included in every other plugin and in (latest) pelican, I assume that is what is wanted.

File: pyproject.toml (hunk 5 above)
Comment (thecaligarmo) [reply pyproject.toml]: Because with the latest version of pytest (5.4.1) there is a bug which preaks with pytest_sugar (see: https://github.com/pytest-dev/pytest/issues/6951) So I put the highest version of pytest (that I could find) that made the plug-in work. Or else the plug-in will stop working if people upgrade pytest.

File: tasks.py (hunk 6 above)
Comment (thecaligarmo) [reply tasks.py:13]: I was just following the standard of all the other plugins and pelican itself. All the other ones are using virtual env and in order to keep everything consistent it would make sense to replicate those no?

File: tasks.py (hunk 7 above)
Comment (thecaligarmo) [reply tasks.py]: All the other ones do... It would make sense to have a unified coding standard acorss the board no?

File: RELEASE.md (hunk 8 above)
Comment (thecaligarmo) [reply RELEASE.md:4]: I followed the guide on Pelican's website. See: https://docs.getpelican.com/en/latest/contribute.html#submitting-your-changes

File: README.md (hunk 1 above)
Comment (thecaligarmo) [reply README.md]: I'll fix this. I put it in the wrong place. Sorry.

File: pelican/plugins/more_categories/more_categories.py (hunk 2 above)
Comment (justinmayer) [reply pelican/plugins/more_categories/more_categories.py]: Once this plugin has been published to PyPI, if the goal is to be able to run `pip install pelican-more-categories` and have everything Just Work™, I believe the next release (tentatively planned as version 4.5) would be required, as the namespace plugin functionality was merged to master after v4.2 was released.

File: pelican/plugins/more_categories/more_categories.py (hunk 3 above)
Comment (justinmayer) [reply pelican/plugins/more_categories/more_categories.py]: While I thought it would be beneficial to have consistent code style across the board, I understand that code style is very much subjective and something that shouldn't be forcibly foisted upon plugin developers who prefer a different style. So plugin authors are free to use their own conventions, as @oulenz has done with More Categories. For example, as you can see, Black is (purposely) not among the tools run during the `lint` task in this repository. So while I fully understand and appreciate your intention here, it would be best to honor this repository's style and not change these quotation marks. :blush:

File: pyproject.toml (hunk 5 above)
Comment (justinmayer) [reply pyproject.toml]: Indeed. See also: https://github.com/Teemu/pytest-sugar/issues/187 That said, rather than hard-pinning, it might be better to instead ensure the problematic versions are not installed, via something like:

     pytest = ">5.0,!=5.4.0,!=5.4.1"

File: tasks.py (hunk 6 above)
Comment (justinmayer) [reply tasks.py:13]: Well, not if it disrupts the preferred tooling of plugin authors and others who prefer alternate workflows. 😉

File: tasks.py (hunk 7 above)
Comment (justinmayer) [reply tasks.py]: Since plugin authors are volunteering their time and effort to maintain plugins, it's best to honor their wishes in this regard. For background, please read: https://github.com/getpelican/cookiecutter-pelican-plugin/issues/2

File: RELEASE.md (hunk 8 above)
Comment (justinmayer) [reply RELEASE.md:4]: That's my fault, Oliver. This bit is related to [AutoPub](http://github.com/autopub/autopub), a tool that I built to automate GitHub + PyPI releases upon pull request merge. I had meant to add AutoPub integration to More Categories as well, but that to-do item seems to have slipped between the cracks. I will submit a separate pull request for that for you to review, which will include continuous integration (CI) for More Categories, so its tests will automatically be run on pull requests such as this one. 😄

File: pyproject.toml (hunk 5 above)
Comment (justinmayer) [reply pyproject.toml]: Since I wanted to get the tests to pass in the CI-related PR I just submitted (#3), I went ahead and made the above modification. So you should back out your change to this line in your PR's branch and rebase on current master.

File: pelican/plugins/more_categories/more_categories.py (hunk 2 above)
Comment (oulenz) [reply pelican/plugins/more_categories/more_categories.py]: @thecaligarmo Changing that to 4.0.0 is fine, absolutely!

@justinmayer Yes, but, users with older versions of pelican could still use this plugin the old-fashioned way right? Also, any particular reason why you want to skip from 4.2 to 4.5, or are you planning intermediate releases?

Comment (justinmayer) [reply pelican/plugins/more_categories/more_categories.py]: @oulenz: Good point. I agree that specifying 4.0.0 makes more sense. Regarding the next release, current master drops Python 2.7 support and will most likely also drop Python 3.5 support. Plus, there are a [decent number of changes in master since 4.2](https://github.com/getpelican/pelican/compare/4.2.0...HEAD). As I mentioned in [another thread](https://github.com/pelican-plugins/webring/issues/2#issuecomment-614599329), there aren't any changes I consider to be truly backwards-incompatible, so 5.0 seemed too big and an unwarranted jump, while 4.3 didn't seem to capture the extent of the changes, so I split the difference and suggested 4.5. Sometimes strict semantic versioning doesn't fit all situations. 🤷‍♂️

File: pelican/plugins/more_categories/more_categories.py (hunk 3 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: Ok cool. I'll go ahead and undo those then. Sorry about that @oulenz

File: tasks.py (hunk 7 above)
Comment (thecaligarmo) [reply tasks.py]: I'll go ahead and undo this.

File: RELEASE.md (hunk 8 above)
Comment (thecaligarmo) [reply RELEASE.md:4]: So should I leave this one as is for now?

File: tasks.py (hunk 6 above)
Comment (thecaligarmo) [reply tasks.py:13]: ok. Sorry .-. I'll go ahead and undo this as well.

Comment (thecaligarmo) [reply tasks.py:13]: So for this one. If I changed it so that it used the same thing like poetry:
```
POETRY = which("poetry") if which("poetry") else (VENV / Path("bin") / "poetry")
```
would that be ok @oulenz ? It would be the ebst of both worlds right? It uses paths when they exist, else it looks inside the environment bin. Or would you prefer I just use the same method that you had before?

Comment (oulenz) [reply tasks.py:13]: I think that would work. I have to admit, I am getting utterly confused by how all these tools inter-operate. In this case, what I don't understand is why we'd want to use invoke to call poetry, rather than the other way around. If you do `poetry run invoke ...`, doesn't poetry then handle the environment, so we could have a cleaner `tasks.py` file? :thinking:

File: pelican/plugins/more_categories/more_categories.py [hunk 9]
```diff
@@ -75,9 +74,24 @@ def create_categories(generator):
...
+            if anc != category:
+                if anc.slug in cats:
+                    cats[anc.slug].append(category)
+                else:
+                    cats[anc.slug] = [category]
+    for category, articles in generator.categories:
+        if category.slug in cats:
+            category.subcategories = cats[category.slug]
+        else:
+            category.subcategories = []
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: Could you split this up into two attributes, `children` and `descendents` (which includes the children)?

Also, please sort them to make things more convenient in the theme templates.

File: pelican/plugins/more_categories/more_categories.py [hunk 10]
```diff
@@ -75,9 +74,24 @@ def create_categories(generator):
...
     generator.categories = sorted(
         list(cat_dct.items()),
-        reverse=generator.settings.get('REVERSE_CATEGORY_ORDER') or False,
+        reverse=generator.settings.get("REVERSE_CATEGORY_ORDER") or False,
     )
-    generator._update_context(['categories'])
+    generator._update_context(["categories"])
+
+    # Add subcategories
+    cats = {}
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: If you initialise this as `cats = defaultdict(list)`, you don't need to test whether it already contains a given category on L86.

File: pelican/plugins/more_categories/more_categories.py [hunk 11]
```diff
@@ -75,9 +74,24 @@ def create_categories(generator):
...
+    for category, articles in generator.categories:
+        for anc in category.ancestors:
+            if anc != category:
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: If you iterate from `categorate.ancestors[1:]`, you don't need this check.

File: pelican/plugins/more_categories/more_categories.py [hunk 12]
```diff
@@ -75,9 +74,24 @@ def create_categories(generator):
...
+                if anc.slug in cats:
+                    cats[anc.slug].append(category)
+                else:
+                    cats[anc.slug] = [category]
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: You can simply use the category itself as the key, since its hash is the hash of its slug.

**Review by oulenz — CHANGES_REQUESTED** [review]

File: pelican/plugins/more_categories/more_categories.py (hunk 9 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: For sure. Do you care which method of sort? Or just the normal `sort()` method on lists?

File: tasks.py (hunk 6 above)
Comment (thecaligarmo) [reply tasks.py:13]: Ok, I'll go ahead and do that for now =) Maybe @justinmayer  would have more idea why we use invoke for something like this? I'm still a newbie... .-.

Comment (justinmayer) [reply tasks.py:13]: My task automation priorities have been to (1) be flexible and accommodate different workflows and (2) make it easy for folks to set up a Pelican development environment. Regarding the first priority, there are (as we've discussed) many ways to handle virtual environments. I prefer to manage mine manually, rather than let Poetry manage them for me. At the same time, people should also be free to let Poetry handle virtual environments on their behalf. Along similar lines, some people (like me) prefer having Poetry and Pre-commit installed globally instead of having them only live in virtual environments. Some folks don't.

Regarding the second objective, the reason I use Invoke to call Poetry here is to keep things simple for new contributors. By putting as much as possible into Invoke tasks, that means that installing Invoke becomes the primary pre-requisite, and then Invoke can handle most of the rest. So in theory, `pip install invoke && invoke setup` becomes all that's required to get going. The concept of letting Poetry manage the environment and then using `poetry run invoke …` is valid enough, but that would depend on Poetry being globally installed, and I didn't want to force users to globally install anything just in order to contribute to the project. The way I configured tasks.py and the [relevant instructions](http://docs.getpelican.com/en/latest/contribute.html#setting-up-the-development-environment) is that it all works with a minimal number of steps and regardless of whether Poetry and Pre-commit are globally installed or not.

**Comment by thecaligarmo** [comment]
@oulenz Ok, I think I fixed everything you wanted and (hopefully) properly reverted the files to their original states in regards to coding style. Let me know if I missed anything.

File: pelican/plugins/more_categories/more_categories.py [hunk 13]
```diff
@@ -79,6 +78,24 @@ def create_categories(generator):
...
     generator._update_context(['categories'])
 
+    # Add subcategories and children
+    descendents = defaultdict(list)
+    children = defaultdict(list)
+    for category, articles in generator.categories:
+        for anc in category.ancestors[1:]:
+            descendents[anc].append(category)
+            if anc == category.parent:
+                children[anc].append(category)
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: This is slightly inelegant, could you place it outside the inner loop, and change it to:

```
if category.parent:
    children[category.parent].append(category)
```

File: pelican/plugins/more_categories/more_categories.py [hunk 14]
```diff
@@ -79,6 +78,24 @@ def create_categories(generator):
...
+    for category, articles in generator.categories:
+        if category.slug in descendents:
+            category.descendents = descendents[category]
+        else:
+            category.descendents = []
+        if category.slug in children:
+            category.children = children[category]
+        else:
+            category.children = []
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: Since we're using defaultdicts, we can simply do:

```
category.descendents = sorted(descendents[category])
category.children = sorted(children[category])

File: pelican/plugins/more_categories/more_categories.py [hunk 15]
```diff
@@ -79,6 +78,24 @@ def create_categories(generator):
     )
     generator._update_context(['categories'])
 
+    # Add subcategories and children
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: Change to `children and descendents`

File: README.md [hunk 16]
```diff
@@ -47,6 +47,13 @@ case this plugin in present, use:
 Likewise, `category.shortname`, `category.parent` and `category.ancestors` can
 also be used on the category template.
 
+Additionally, this plugin adds `category.subcategories`: a `list` of categories
+that have `category` as a parent.
+
+    {% for subcat in category.subcategories %}
+        <a href="{{ SITEURL }}/{{subcat.url}}">{{subcat.shortname|capitalize}}</a>
+    {% endfor %}
+
```
Comment (oulenz) [inline README.md]: "`category.children` and `category.descendents`", and change the example to use `category.children`

File: tasks.py (hunk 6 above)
Comment (oulenz) [reply tasks.py:13]: Thanks for the explanation, it's helpful to understand what is logically necessary and what are design decisions. I agree that focusing everything on invoke and `tasks.py` is probably the easiest to understand for new contributors.

File: tasks.py [hunk 17]
```diff
@@ -1,12 +1,25 @@
...
+VENV = str(VENV_PATH.expanduser())
+
+
+TOOLS = ['poetry', 'pre-commit']
+FLAKE8 = which('flake8') if which('flake8') else (VENV / Path('bin') / 'flake8')
+ISORT = which('isort') if which('isort') else (VENV / Path('bin') / 'isort')
+PYTEST = which('pytest') if which('pytest') else (VENV / Path('bin') / 'pytest')
+PIP = which('pip') if which('pip') else (VENV / Path('bin') / 'pip')
+POETRY = which('poetry') if which('poetry') else (VENV / Path('bin') / 'poetry')
+PRECOMMIT = which('pre-commit') if which('pre-commit') else (VENV / Path('bin') / 'pre-commit')
```
Comment (oulenz) [inline tasks.py]: I think these ternary statements can all be simplified to:

```
FLAKE8 = which('flake8') or Path(f'{VENV}/bin/flake8')
```

Also, TOOLS should come at the end, and say

```
TOOLS = [POETRY, PRECOMMIT]
```

**Review by oulenz — CHANGES_REQUESTED** [review]

File: pelican/plugins/more_categories/more_categories.py (hunk 13 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: That's much better. Good idea =D

File: pelican/plugins/more_categories/more_categories.py (hunk 14 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: I'm not used to defaultdicts .-. I really need to get used to them though cause that's so much nicer. =D

File: tasks.py (hunk 17 above)
Comment (thecaligarmo) [reply tasks.py]: If we change `TOOLS` to what you have, won't that break the loop that iterates over `TOOLS` in `tools()`? It feels counter-productive to do `TOOLS = [POETRY, PRECOMMIT]` since the point of `TOOLS` is to install `poetry` and `pre-commit` if they are not already installed.

Comment (oulenz) [reply tasks.py]: That's a good point! So if `PIP` is set to `Path(f'{VENV}/bin/pip')`, and we run `f'{PIP} install {tool}'`, will the tool be automatically installed in the virtual_env?

Comment (thecaligarmo) [reply tasks.py]: From what I understand, it should.

File: pelican/plugins/more_categories/more_categories.py [hunk 18]
```diff
@@ -79,6 +78,18 @@ def create_categories(generator):
     )
     generator._update_context(['categories'])
 
+    # Add descendents and children
+    descendents = defaultdict(list)
+    children = defaultdict(list)
+    for category, articles in generator.categories:
+        for anc in category.ancestors[1:]:
```
Comment (oulenz) [inline pelican/plugins/more_categories/more_categories.py]: Sorry, this should be `category.ancestors[:-1]`

**Review by oulenz — CHANGES_REQUESTED** [review]
I've found one remaining issue (which I suggested, sorry!).

Other than that, could you rearrange all commits into three commits?

* one with the configuration changes
* one with the added functionality and related documentation changes
* one with the `RELEASE.MD` file

If you want, I can also do this for you.

In addition, I've written some tests to check ancestors/descendents functionality, which I'll add.

File: pelican/plugins/more_categories/more_categories.py (hunk 18 above)
Comment (thecaligarmo) [reply pelican/plugins/more_categories/more_categories.py]: Fixed =)

Also, for the "rearrange all commits into three commits" part. To be honest, I'm not sure how to do that .-. Could you do that? (Or tell me what I should be doing and I can do it)

**Comment by oulenz** [comment]
@thecaligarmo I've rearranged your commits and submitted them via a new pull request. (I've taken care so you're still listed as the author.)

The idea in general is that commits should be relatively self-contained, so that if you are searching through the commit history, it's easier to understand when and why something was changed. It is natural to get feedback on pull request, which leads to new commits. These can be combined by squashing/amending (though not necessarily to just one commit per pull request, there can be more if several things were changed). In addition, it can happen that the changes in your pull request clash with changes added to the master in the meantime, which requires a rebase.

In this case, it took some interactive rebasing to untangle the changes in your commits, so I went ahead and did it myself (also because I haven't used interactive rebasing much before, and I enjoyed the challenge!)

Anyway, many thanks for your contribution!
